### PR TITLE
feat: account for category hierarchy in loss

### PR DESCRIPTION
**Pull request by alexgarel**
Loss should account for taxonomy hierarchy

### CODE REVIEW

File: lib/taxonomy.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        scats = set(cats)
+        for cat in cats:
+            scats.update(self.ancestors(cat))
+        return scats
+
+    def cats_compat(self, cats):
+        """Define the compatible and incompatible categories for a list of categories
+
+        The idea is that we don't want to penalize categories that are found by the model,
+        because it goes deeper than the
```
Comment (AnisZakari) [inline lib/taxonomy.py:42]: some words are missing. Maybe you meant "we don't want to penalize deeper categories found by the model"

File: requirements.txt [hunk 2]
```diff
@@ -4,10 +4,11 @@ cython==0.29.28
 dacite==1.6.0
 graphviz==0.20
 matplotlib==3.4.3
+networkx==2.6.3
 numpy==1.19.5
 pandas==1.3.5
 pydot==1.4.2
 robotoff==1.5.0
 scikit-learn==1.0.2
-tensorflow==2.6.3
```
Comment (AnisZakari) [inline requirements.txt:12]: is there a specific reason to change the tensorflow version ?

**Review by AnisZakari — CHANGES_REQUESTED** [review]
As the lib eval was removed and as there's no more serving and training model  there are some changes needed in the notebooks. 

There is also a problematic behaviour of the model in the prediction process: 
- the model expects a `compat` key in `ds_test`. 
- we don't want the model to apply masks on the outputs during the prediction process.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 3]
```diff
@@ -0,0 +1,1119 @@
...
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import tensorflow as tf\n",
+    "import tensorflow_datasets as tfds\n",
+    "from tensorflow.keras import callbacks, layers\n",
+    "from tensorflow.keras.utils import plot_model\n",
+    "\n",
+    "from lib.dataset import *\n",
+    "from lib.directories import init_cache_dir, init_model_dir\n",
+    "from lib.eval import top_predictions\n",
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:140]: yep I have updated whole notebook… after rebasing.

File: lib/taxonomy.py (hunk 1 above)
Comment (alexgarel) [reply lib/taxonomy.py:42]: well spotted

```
        The idea is that we don't want to penalize categories that are found by the model,
        because it goes deeper than what the dataset is providing
        (database is not necessarily complete)
```

File: requirements.txt (hunk 2 above)
Comment (alexgarel) [reply requirements.txt:12]: As it is a minor version, I think it is ok.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 4]
```diff
@@ -0,0 +1,1119 @@
...
+    "    super().__init__(*args, **kwargs)\n",
+    "    self.mask_feature = mask_feature\n",
+    "\n",
+    "  def compute_loss(self, x=None, y=None, y_pred=None, sample_weight=None):\n",
+    "    # first mask categories we do not want using the compat feature\n",
+    "    mask = x[self.mask_feature]\n",
+    "    # zeros mask_feature in y_pred, as we know they are not in y_pred\n",
+    "    y_pred *= mask\n",
+    "    return super().compute_loss(x, y, y_pred, sample_weight)"
+   ]
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:643]: I committed the version using it.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 5]
```diff
@@ -0,0 +1,1119 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7d6742be",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pathlib import Path\n",
+    "MODEL_DIR =  Path(\"../model.20220610-094814\")"
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:926]: thanks, updated from master, there also.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 6]
```diff
@@ -0,0 +1,1119 @@
...
+   "id": "2df9fcc3",
+   "metadata": {
+    "id": "2df9fcc3",
+    "outputId": "ae0b7652-98b7-44df-d2f2-6367e21a1743"
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "serving_pred_test = serving_model.predict(ds_test.padded_batch(128))\n",
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:1045]: updated, thanks.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 7]
```diff
@@ -0,0 +1,1119 @@
...
+   "id": "554e2998",
+   "metadata": {
+    "id": "554e2998",
+    "outputId": "e59ae90d-f414-4768-e5b5-6c51887ba42e"
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "test_top_predictions = top_predictions(ds_test, serving_pred_test, 4)"
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:1062]: updated, thanks.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 8]
```diff
@@ -0,0 +1,1119 @@
...
+   "id": "fcb0f069",
+   "metadata": {
+    "id": "fcb0f069",
+    "outputId": "18cca2bd-3af2-4244-9c70-63b4989443c7"
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "training_pred_test = training_model.predict(ds_test.padded_batch(128))\n",
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:1017]: Resolved by using MaskingModel instead.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 9]
```diff
@@ -0,0 +1,1119 @@
...
+    "          )\n",
+    "        return (x, y)\n",
+    "\n",
+    "    # applies to non-batched dataset\n",
+    "    return (\n",
+    "        ds\n",
+    "        .map(_transform, num_parallel_calls=tf.data.AUTOTUNE, deterministic=True)\n",
+    "        .apply(filter_empty_labels)\n",
+    "    )\n"
+   ]
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:609]: Refactoring to take this into accountt.

File: experiments/Train-with-cat-compat-loss.ipynb [hunk 10]
```diff
@@ -0,0 +1,1119 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cff4364b",
+   "metadata": {
+    "id": "cff4364b"
+   },
+   "outputs": [],
+   "source": [
+    "ds_test = load_dataset('off_categories', split=TEST_SPLIT)"
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply experiments/Train-with-cat-compat-loss.ipynb:992]: resolved using MaskingModel.

**Review by AnisZakari — APPROVED** [review]
LGTM !
